[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112078
* #111726
* #111725
* #111415
* #111614
* #111717
* #111306
* #112122

[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

voznesenskym committed:
[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

[ghstack-poisoned]

jansel commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -587,11 +586,7 @@ def repl(v: VariableTracker):
-                (isinstance(x, LazyMutableLocal) and x.vt is not None)
-                for x in v.recursively_contains
-            )
+            return oldvar.mutable_local not in v.recursively_contains
```
I think we still need something like this.  It is needed for correctness.

voznesenskym replied:
Hmm, ok. I wasn't sure where this was needed - since this is just for update_locals_and_stack and accessing mutable_local realizes it, it felt always sound to just let the update_locals_and_stack step realize all the candidates. Is that not effectively the same?

jansel replied:
Made obsolete by https://github.com/pytorch/pytorch/pull/112122

voznesenskym replied:
Sgtm

jansel committed:
Update on "[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container"


[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container"


[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container"


[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel committed:
Update on "[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container"


[dynamo] Replace mutable_local for LazyVariableTracker with a realizing property and separate container

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
